ESQL: Honor skip_unavailable setting for nonmatching indices errors at planning time

quux00 opened the pull request:
Adds support to ES|QL planning time (EsqlSession code) for dealing with non-matching indices and
how that relates to the remote cluster skip_unavailable setting and also how to deal with missing
indices on the local cluster (if included in the query).

For clusters included in an ES|QL query:

• `skip_unavailable=true` means:
  if no data is returned from that cluster (due to cluster-not-connected, no matching indices, a missing
  concrete index or shard failures during searches), it is not a "fatal" error that causes the entire query to fail. 
  Instead it is just a failure on that particular cluster and partial data should be returned from other clusters.

• `skip_unavailable=false` means: 
   if no data is returned from that cluster (for the same reasons enumerated above), then the whole query
should fail. This allows users to ensure that data is returned from a "required" cluster.					

• For the local cluster, ES|QL assumes `allow_no_indices=true` and the skip_unavailable setting does not apply
  (in part because there is no way for a user to set skip_unavailable for the local cluster)

Based on discussions with ES|QL team members, we defined the following rules to be enforced with respect to non-matching index expressions:

**Rules enforced at planning time**

P1. fail the query if there are no matching indices on any cluster (VerificationException)
P2. fail the query if a skip_unavailable:false cluster has no matching indices (the local cluster already has this rule enforced at planning time)
P3. fail query if the local cluster has no matching indices and a concrete index was specified

**Rules enforced at execution time**

For missing concrete (no wildcards present) index expressions:
E1. fail the query when it was specified for the local cluster or a skip_unavailable=false remote cluster
E2: on skip_unavailable=true clusters: an error fails the query on that cluster, but not the entire query
    (data from other clusters still returned)

**Notes on the rules**

P1: this already happens, no new code needed in this PR
P2: The reason we need to enforce rule 2 at planning time is that when there are no matching indices from
  field caps the EsIndex that is created (and passed into IndexResolution.valid) leaves that cluster out of the
  list, so at execution time it will not attempt to query that cluster at all, so execution time will not catch
  missing concrete indices. And even if it did get queried at execution time it wouldn't fail on wildcard only
  indices where none of them matched.
P3: Right now `FROM remote:existent,nomatch` does NOT throw a failure (for same reason described in rule 2 above)
so that needs to be enforced in this PR.

This PR deals with enforcing and testing the planning time rules: P1, P2 and P3. A follow-on PR will address changes needed for handling the execution time rules.

**Notes on PR scope**

This PR covers nonsecured clusters (`xpack.security.enabled: false`) and security using certs ("RCS1). 
In my testing I've founding that api-key based security ("RCS2") is not behaving as expected, so that
work has been deferred to a follow-on PR.

Partially addresses https://github.com/elastic/elasticsearch/issues/114531

quux00 committed:
Intmd commit - going to split into refactoring ticket

quux00 committed:
updateExecutionInfoWithClustersWithNoMatchingIndices updated to follow new rules; tests updated

elasticsearchmachine commented:
Hi @quux00, I've created a changelog YAML for you.

quux00 committed:
Update docs/changelog/116348.yaml

quux00 committed:
Fixed checkstyle issues

smalyshev commented on x-pack/plugin/esql/src/internalClusterTest/java/org/elasticsearch/xpack/esql/action/CrossClustersEnrichIT.java:
```diff
@@ -61,6 +61,8 @@
 import static org.hamcrest.Matchers.equalTo;
 import static org.hamcrest.Matchers.greaterThanOrEqualTo;
 
+// MP TODO: what to do about missing enrich policies on skip_unavailable=true clusters? Should that be fatal or non-fatal?
```
I'd say they'd be SKIPPED too. If we are skipping on missing index, it makes sense treat the enrich policy the same way, especially given the potential move from enrich policies to indices in the future.

quux00 replied:
After looking at this more closely, I don't see how to cleanly/easily handle missing enrich policies from skip_unavailable=true clusters. We would need to add knowledge to the Analyzer and LogicalPlanner about whether to not have that be a fatal error and then actively rewrite the users query to remove the offending cluster from the query entirely.

That's certainly beyond what I want to do in this PR and possibly not something we ever want to do. Open to feedback on it from the ES|QL team.

quux00 replied:
Scratch that. I believe I've found a way to deal with missing enrich policies for skip_unavailable=true clusters without having to modify code in/around the LogicalPlan. Need to do some extensive testing on it, so that can either be a later commit to this PR or a follow-on PR.

smalyshev replied:
If I understood it correctly (very well could be not) the enrich resolution step is supposed to prepare a map where for each cluster we have a mapping that maps the policy to specific enrich indices, etc. Now, if we encounter the situation where, in remote enrich (in local-side enrich the local one should be used, as I understand) the policy is missing, and we mark the cluster as skipped, then nothing will be run on that cluster. So the missing map entry for this cluster should be ok? I _think_ the planner should be ok with that too, but we will need to ensure the verifier wouldn't give any trouble for skipped clusters. I'm not sure why query rewriting is necessary, would be glad to sync on this, I may be missing something.

smalyshev commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSessionCCSUtils.java:
```diff
@@ -191,29 +189,83 @@ static void updateExecutionInfoWithClustersWithNoMatchingIndices(EsqlExecutionIn
+
+    // visible for testing
+    static boolean missingIndicesIsFatal(String clusterAlias, EsqlExecutionInfo executionInfo) {
+        if (executionInfo.getCluster(clusterAlias).isSkipUnavailable()) {
```
Why not `executionInfo.isSkipUnavailable`?

quux00 replied:
In production code they are the same so it doesn't matter too much, but in testing code it is easier to just set the value on the Cluster object rather than to also have to set it up in the Function passed to EsqlExecutionInfo. 

Once it is set on the EsqlExecutionInfo.Cluster object, IMO it is generally preferable to grab it from the Cluster object rather than re-inquire it from the RemoteClusterService. But it is a minor nitpick issue.

Ideally we wouldn't have the EsqlExecutionInfo.isSkipUnavailable method at all, but rather grab that (initially) from the RemoteClusterService and set it once on each Cluster, but the ESQL code (unlike the _search code) was not amenable to that so I went with the idea of getting a reference to the RemoteClusterService functionality via the EsqlExecutionInfo.isSkipUnavailable method.

quux00 committed:
It turns out we cannot rely on Map<String, IndexMode> indexNameWithModes in EsIndex to always have the list
of indices resolved by field-caps because when there are no mappings found in that index we pass in an empty
map (which seems wrong). I tried to stop doing that but soon hit layers of ESQL I don't understand, so tried
the simplier approach of adding a new (somewhat redundant) new field to IndexResolver: Set<String> resolvedIndices.

This solves the problem of allowing the CCS handler code in EsqlSession.preAnalyze (which calls
EsqlSessionCCSUtils.updateExecutionInfoWithClustersWithNoMatchingIndices) to determine whether or not a cluster
had no matching indices from the field-caps call, allowing either an error to be thrown or for the CCS metadata to
be updated.

quux00 committed:
Fixed forbiddenAPI issues in CrossClustersQueryIT

quux00 committed:
Added CrossClusterEsqlRCS1MissingIndicesIT

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql-ccs/skip_unavailable-missing-indices-t2

quux00 committed:
Added index aliases and filtered alias to CrossClustersQueryIT for nonmatching indices tests

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql-ccs/skip_unavailable-missing-indices-t2

quux00 committed:
Added tests against index with no mappings to ensure correct error message is returned

Adjusted RemoteClusterSecurityEsqlIT so it passes for skip_unavailable=true, although
the error message returned for skip_unavailable=false is still wrong.

quux00 committed:
I have decided that this ticket will NOT try to handle missing enrich policies for skip_un=true clusters; requires modifying the Analyzer

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql-ccs/skip_unavailable-missing-indices-t2

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql-ccs/skip_unavailable-missing-indices-t2

smalyshev commented on docs/changelog/116348.yaml:
```diff
@@ -0,0 +1,6 @@
+pr: 116348
+summary: "DRAFT: ESQL honors `skip_unavailable` setting for nonmatching indices errors\
```
We can drop "DRAFT" here I think?

smalyshev commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSessionCCSUtils.java:
```diff
@@ -181,39 +180,92 @@ static void updateExecutionInfoWithUnavailableClusters(EsqlExecutionInfo execInf
+            return false;
+        }
+        // missing indices on local cluster is fatal only if a concrete index requested
+        if (clusterAlias.equals(RemoteClusterAware.LOCAL_CLUSTER_GROUP_KEY)) {
```
Doesn't it make sense to reverse the order of if's - if we're on local cluster, no point to check `isSkipUnavailable` first, it'll never be true?

quux00 replied:
Good idea. And that allows me to simplify the logic to:

```
        if (clusterAlias.equals(RemoteClusterAware.LOCAL_CLUSTER_GROUP_KEY)) {
            return concreteIndexRequested(executionInfo.getCluster(clusterAlias).getIndexExpression());
        }
        return executionInfo.getCluster(clusterAlias).isSkipUnavailable() == false;
```

so one fewer "if" check.
Fixed in next push.

quux00 committed:
PR feedback - simplified EsqlSessionCCSUtils.missingIndicesIsFatal

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql-ccs/skip_unavailable-missing-indices-t2

pawankartik-elastic commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSessionCCSUtils.java:
```diff
@@ -181,41 +180,90 @@ static void updateExecutionInfoWithUnavailableClusters(EsqlExecutionInfo execInf
+    }
+
+    private static boolean concreteIndexRequested(String indexExpression) {
+        for (String expr : indexExpression.split(",")) {
```
Do we really have to split by `,` here? What happens if we directly loop over `indexExpression` and check for `*`?

quux00 replied:
We are looping over indexExpression - you have to split it out into the separate index expressions to check.

`foo*,myremote:foo,myremote:bar*` has a concrete index in it, but you can't just do indexExpression.indexOf("*") to check, you have to check each `foo*`,  `myremote:foo` and `myremote:bar*` separately.

pawankartik-elastic replied:
That makes sense. For some reason, I was under the assumption that the method returns `true` if all indices are concrete.

quux00 committed:
Now handling date math in EsqlSessionCCSUtils.concreteIndexRequested

quux00 committed:
Updatec yaml changelog with proper title and issue ref

nik9000 commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/VerifierTests.java:
```diff
@@ -86,7 +86,8 @@ public void testUnsupportedAndMultiTypedFields() {
                 Map.of(unsupported, unsupportedField, multiTyped, multiTypedField, "int", unsupportedField, "double", multiTypedField)
-            )
+            ),
+            ipIndices
```
Should this have all six indices in it? Maybe it doesn't matter, but more realistic I think.

nik9000 reviewed (commented):
Oh boy! It looks right. I left a tiny thing. I'm going to give it a little rest in my head for a bit and come back soon and look again and approve.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/index/IndexResolution.java:
```diff
@@ -12,22 +12,28 @@
 
-    public static IndexResolution valid(EsIndex index) {
-        return valid(index, Collections.emptyMap());
+    public static IndexResolution valid(EsIndex index, Set<String> resolvedIndices) {
```
Nit: to avoid the noise in the PR, keep the old method in place and extract the name from EsIndex:
```
valid(EsIndex index) { return valid(index,  singletonSet(index.name())}
```

quux00 replied:
Good idea. Except we should use index.concreteIndices(), not index name, since the latter is the original unresolved user requested index expression. Fixed in next push.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/index/IndexResolution.java:
```diff
@@ -79,6 +93,13 @@ public Map<String, FieldCapabilitiesFailure> getUnavailableClusters() {
+    /**
+     * @return all indices found by field-caps (regardless of whether they had any mappings)
+     */
+    public Set<String> getResolvedIndices() {
```
Nit: since there's no setter, you can drop the get -> resolvedIndices()

quux00 replied:
OK. Fixed in next push. I also changed `getUnavailableClusters()` to `unavailableClusters()`.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/index/IndexResolution.java:
```diff
@@ -87,12 +108,13 @@ public boolean equals(Object obj) {
+        return Objects.hash(index, invalid, resolvedIndices, unavailableClusters);
     }
 
     @Override
```
You probably want to have the resolved indices in toString()

quux00 replied:
Good idea. Updated in next push.

costin reviewed (commented):
Minor drive-by comments.

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql-ccs/skip_unavailable-missing-indices-t2

quux00 committed:
PR feedback: Removed IndexResolution.valid(EsIndex index,Set<String> resolvedIndices) to avoid PR noise, since it can be inferred from EsIndex contents

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql-ccs/skip_unavailable-missing-indices-t2

quux00 commented:
@elasticsearchmachine run elasticsearch-ci/part-1

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql-ccs/skip_unavailable-missing-indices-t2

quux00 committed:
Update docs/changelog/116348.yaml

elasticsearchmachine commented:
Hi @quux00, I've updated the changelog YAML for you.

quux00 committed:
Changing IndexResolution toString for invalid cases causes tests to break, so reverting original model for invalid case

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql-ccs/skip_unavailable-missing-indices-t2

quux00 committed:
Fix changelong error

elasticsearchmachine commented:
## 💚 Backport successful
| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [8.x](https://github.com/elastic/elasticsearch/pull/116824)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/116824">](https://github.com/elastic/elasticsearch/pull/116824) |
